src/mlpack/methods/mean_shift/mean_shift_impl.hpp

```diff
@@ -0,0 +1,129 @@
+/**
+ * @file mean_shift_impl.hpp
+ * @author Shangtong Zhang
+ *
+ * Mean Shift clustering
+ */
+
+#include "mean_shift.hpp"
+#include <mlpack/core/kernels/gaussian_kernel.hpp>
+
+namespace mlpack {
+namespace meanshift {
+  
+/**
+  * Construct the Mean Shift object.
+  */
+template<typename KernelType,
+         typename MatType>
+MeanShift<
+  KernelType,
+  MatType>::
+MeanShift(const double duplicateThresh,
+          const size_t maxIterations,
+          const double stopThresh,
+          const KernelType kernel) :
+    duplicateThresh(duplicateThresh),
+    maxIterations(maxIterations),
+    stopThresh(stopThresh),
+    kernel(kernel)
+{
+  // Nothing to do.
+}
+
+/**
+  * Perform Mean Shift clustering on the data, returning a list of cluster
+  * assignments and centroids.
+  */
+template<typename KernelType,
+         typename MatType>
+inline void MeanShift<
+    KernelType,
+    MatType>::
+Cluster(const MatType& data,
+        arma::Col<size_t>& assignments,
+        arma::mat& centroids) {
+  
+  // all centroids before remove duplicate ones.
+  arma::mat allCentroids(data.n_rows, data.n_cols);
+  assignments.set_size(data.n_cols);
+  
+  // for each point in dataset, perform mean shift algorithm.
+  for (size_t i = 0; i < data.n_cols; ++i) {
+    
+    size_t completedIterations = 0;
+    
+    //initial centroid is the point itself.
+    allCentroids.col(i) = data.col(i);
+    
+    while (true) {
+      
+      // new centroid
+      arma::Col<double> newCentroid = arma::zeros(data.n_rows, 1);
+      
+      double sumWeight = 0;
+      for (size_t j = 0; j < data.n_cols; ++j) {
+        
+        // calc weight for each point
+        double weight = kernel.Evaluate(allCentroids.col(i), data.col(j));
+        sumWeight += weight;
+        
+        // update new centroid.
+        newCentroid += weight * data.col(j);
+        
+      }
+      
+      newCentroid /= sumWeight;
+      
+      completedIterations ++;
+      
+      // calc the mean shift vector.
+      arma::Col<double> mhVector = newCentroid - allCentroids.col(i);
+      
+      // update the centroid.
+      allCentroids.col(i) = newCentroid;
+      
+      if (arma::norm(mhVector, 2) < stopThresh ||
+          completedIterations > maxIterations) {
+        break;
+      }
+      
+    }
+    
+  }
+  
+  // remove duplicate centroids.
+  for (size_t i = 0; i < allCentroids.n_cols; ++i) {
```

Comment:
Why not check if there is a duplicate centroid once you finish finding the centroid for an individual point?  If you refactor the code to only consider one centroid at a time and only add it to the results if it is not duplicated in the existing list of results, I think you will get faster (and more memory-efficient) code.